src/shell/plugins/pals.c

```diff
@@ -0,0 +1,493 @@
+/************************************************************\
+ * Copyright 2021 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <flux/hostlist.h>
+#include <flux/shell.h>
+
+#include "jansson.h"
+
+#include "jobspec.h"
+
+/*  PALS --- interface with HPE/Cray's PMI.
+ *
+ * To support Cray PMI, the launcher must perform these tasks:
```

Comment:
Some general comments:
* perhaps the actual libpals could be used to write tests for this?
* maybe include cray in the name?  Like `cray-pals`?  Otherwise the name doesn't provide much useful context.
* a bit more description in the header about the overall architecture would be useful, since usually PMI is a service provided by the launcher; in this case it is a proprietary service that we are just handing off job information to (as I understand it!)
* it would possibly be good to push back on cray on the naming of the port variable since a) env vars beginning with `PMI_` are used by PMI and this is operating at a different level, and b) apparently it's going to be  comma-separated list of ports, so it should be plural.
* seems like it may be a good idea to locate this in the other repo as suggested by @SteVwonder.  That would allow you to update it frequently during early deployment, without requiring a flux-core release.

Reply:
Thanks for the comments, I agree with all of them. Cray would definitely have the upper hand in the environment variable argument because they're already [built into Slurm](https://github.com/SchedMD/slurm/blob/6cb01bd89a92c7eaa80624891c034b2af85727d4/src/plugins/mpi/cray_shasta/mpi_cray_shasta.c#L171) though.